guppylang/compiler/core.py

```diff
@@ -48,6 +48,19 @@ def _compile(self, defn: CheckedDef) -> CompiledDef:
             return defn.compile_outer(self.module)
         return defn
 
+    def compile(self, defn: CheckedDef) -> None:
+        """Compiles the given definition and all of its dependencies into the current
+        Hugr."""
+        if defn.id in self.compiled:
+            return
+
+        self.compiled[defn.id] = self._compile(defn)
+        self.worklist.add(defn.id)
+        while self.worklist:
+            next_id = self.worklist.pop()
+            next_def = self[next_id]
```

Comment:
is there still compilation happening inside `__getitem__`? I don't see any change to that behaviour here

Reply:
Kind of. The `__getitem__` call only builds the empty function def and registers it in the worklist. Imo that's fine, but I can turn it into an explicit method call if you prefer?

Reply:
yes please, mutating __getitem__ is best avoided if possible